src/test/CustomCurveHook.sol

```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.20;
+
+import {Hooks} from "../libraries/Hooks.sol";
+import {SafeCast} from "../libraries/SafeCast.sol";
+import {IHooks} from "../interfaces/IHooks.sol";
+import {IPoolManager} from "../interfaces/IPoolManager.sol";
+import {PoolKey} from "../types/PoolKey.sol";
+import {BalanceDelta, toBalanceDelta} from "../types/BalanceDelta.sol";
+import {Currency} from "../types/Currency.sol";
+import {CurrencySettleTake} from "../libraries/CurrencySettleTake.sol";
+import {BaseTestHooks} from "./BaseTestHooks.sol";
+import {IERC20Minimal} from "../interfaces/external/IERC20Minimal.sol";
+import {CurrencyLibrary, Currency} from "../types/Currency.sol";
+
+contract CustomCurveHook is BaseTestHooks {
+    using Hooks for IHooks;
+    using CurrencyLibrary for Currency;
+    using CurrencySettleTake for Currency;
+
+    error AddLiquidityDirectToHook();
+
+    IPoolManager immutable manager;
+
+    constructor(IPoolManager _manager) {
+        manager = _manager;
+    }
+
+    modifier onlyPoolManager() {
+        require(msg.sender == address(manager));
+        _;
+    }
+
+    function beforeSwap(
+        address, /* sender **/
+        PoolKey calldata key,
+        IPoolManager.SwapParams calldata params,
+        bytes calldata /* hookData **/
+    ) external override onlyPoolManager returns (bytes4, int128) {
+        (Currency inputCurrency, Currency outputCurrency, uint256 amount) = _getInputOutputAndAmount(key, params);
+
+        // this "custom curve" is a line, 1-1
+        // take the full input amount, and give the full output amount
+        manager.take(inputCurrency, address(this), amount);
+        outputCurrency.settle(manager, address(this), amount, false);
```

Comment:
ah nice so this is why we added transfer not just transferFrom to currencySettleTake lib?